src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -40,6 +40,45 @@
 
 namespace knowhere {
 
+__global__ void
+filter(const int k1, const int k2, const int nq, const uint8_t* bs, int64_t* ids_before, float* dis_before,
+       int64_t* ids, float* dis) {
+    int tx = blockIdx.x * blockDim.x + threadIdx.x;
+    int ty = blockIdx.y * blockDim.y + threadIdx.y;
+    extern __shared__ char s[];
+    int64_t* ids_ = (int64_t*)s;
+    float* dis_ = (float*)&s[k2 * sizeof(int64_t)];
+    if (tx >= k2)
+        return;
+    int64_t i = ids_before[ty * k2 + tx];
+    float d = dis_before[ty * k2 + tx];
+    bool check = bs[i >> 3] & (0x1 << (i & 0x7));
+    if (!check) {
+        ids_[tx] = i;
+        dis_[tx] = d;
+    } else {
+        ids_[tx] = -1;
+        dis_[tx] = -1.0f;
+    }
+    __syncthreads();
+    if (tx == 0) {
+        int j = 0, k = 0;
+        while (j < k1 && k < k2) {
+            while (ids_[k] == -1) k++;
+            if (k >= k2)
```

Comment:
low gpu utilization, just one worker thread here. Can be optimized with 'parallel prefix sum' (https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda) or 'parallel sort'(https://developer.nvidia.com/gpugems/gpugems2/part-vi-simulation-and-numerical-algorithms/chapter-46-improved-gpu-sorting)